Add support for async drivers

dellisd opened the pull request:
Regarding #1942

This adds the main codegen changes along with a minimal runtime needed to support async driver implementations, along with some basic (and incomplete) implementations of an R2DBC driver and a driver for running sql.js in a web worker. There's still a lot of work (and cleanup) to be done on it, but I'm hoping to get some feedback on the design of the runtime before going any further.

To summarize how it works, async code generation is enabled in the gradle config:

```kotlin
sqldelight {
  database("MyDatabase") {
    packageName = "com.test.db"
    generateAsync = true
  }
}
```

SQLdelight will then generate queries (and related code) that return an `AsyncSqlDriver.Callback<R>` object that can be used to observe the result/errors. This object can be easily adapted to other reactive/async frameworks, like [these coroutine extensions](https://github.com/dellisd/sqldelight/blob/7eb9012b1446953dba7ba2cfbfb5577dec847966/extensions/coroutines-extensions/src/commonMain/kotlin/app/cash/sqldelight/coroutines/AsyncExtensions.kt).

Drivers implement the `AsyncSqlDriver` interface which is pretty similar to `SqlDriver` except most of the methods now return the callback  object.

## Open Questions

1. Is it worth implementing and maintaining this whole async runtime just to enable support for other frameworks? It adds a lot of complexity both in the runtime implementation and codegen compared to just generating and implementing suspend functions.
2. What to do about transactions? From reading past issues, it looks like there could be trouble when dealing with multithreading? Also, `transactionWithResult` could become trouble since we would presumably want to be able to support the semantics of whatever reactive library is being used which currently isn't possible (related to #1856)  
e.g. for coroutines we'd want something like  
```kotlin
val list = nameQueries.suspendingTransactionWithResult { 
  nameQueries.insert(/* ... */).await()
  nameQueries.getAll().awaitAsList()
}
```  
3. R2DBC returns results as a stream of rows, rather than a cursor that can be controlled directly and synchronously. This is incompatible with the existing `SqlCursor` interface, and my current workaround in the R2DBC driver implementation is to [read all the rows into a map](https://github.com/dellisd/sqldelight/blob/7eb9012b1446953dba7ba2cfbfb5577dec847966/drivers/r2dbc-driver/src/main/kotlin/app/cash/sqldelight/driver/r2dbc/R2dbcDriver.kt#L28-L30) and then [hand that](https://github.com/dellisd/sqldelight/blob/7eb9012b1446953dba7ba2cfbfb5577dec847966/drivers/r2dbc-driver/src/main/kotlin/app/cash/sqldelight/driver/r2dbc/R2dbcDriver.kt#L173-L193) to a `SqlCursor` implementation. _Not ideal_. Adapting all of the mapper code to follow the asynchronous semantics would require a lot more effort to change the codegen.
4. This isn't really related to "async sqldelight" nor is it really a question, but the R2DBC implementations are weird. The h2 and postgres drivers require statement placeholders to use the `$1, $2` format instead of `?, ?` so they don't work with the current `R2dbcDriver`, only mysql works. Special sqldelight driver implementations could probably work around it? But I also don't know how much demand there is for R2DBC in sqldelight, I mainly implemented it just to help inform the design of the async runtime.

Lastly, I threw together a quick sample project that shows the sql.js worker driver in action: https://github.com/dellisd/sqldelight-js-worker

AlecStrong commented:
Have only done a skim, but this is heading in the right direction.

We should create a new module for the async runtime (`async-runtime` or something like that) that has all these new types, and potentially duplicates some other types like `Schema` that I assume we're reusing right now. I'd like for the two worlds to be pretty separate in terms of importing a public API when you switch to async. The gradle plugin adds the runtime dependency so it can pretty smoothly switch to the async runtime if needed. If there is a huge amount of common ground between the async and sync runtimes, we can create a common module or something for them to share.

1. I don't think its worth maintaining. coroutines is the standard for async in kotlin now, I don't think it would be controversial to have it be what we output, and similarly if people want to convert to their own type it will be possible.
2. In the same vein, I think we just expose suspending functions for transactions. I'm unfamiliar with the issues that'll come up because of that - but maybe theres a world where we just enforce that statements _within_ the transaction happen within the same context, or something like that. I assume thats the problem where you can't have the different statements running on different threads since then they may use different database connections. If we record the context the transaction was started with maybe we can avoid that?
3. Cool with that - since theres already a lot going on lets keep this initial implementation to fleshing out the API and then if we want to revisit things or make improvements we can do that in follow ups.
4. No strong feelings here, the only async driver I've seen requested is jasync which has a coroutines compatible API.

dellisd commented:
I've split the runtime module up to separate the async code and also extracted a few classes into a shared module that I don't think would make sense to duplicate (like `ColumnAdapter`). 

There's also `:runtime:internal` which might be overkill for just a single function, but it's to prevent a shared internal function from being exposed to consumers. I can get rid of it if we don't want the whole extra module just for that, and I can also change the names since `:runtime:runtime` just looks silly

AlecStrong commented:
yep, module structure looks good

AlecStrong commented on sqldelight-compiler/src/main/kotlin/app/cash/sqldelight/core/compiler/SqlDelightCompiler.kt:
```diff
@@ -42,12 +42,13 @@ object SqlDelightCompiler {
     file: SqlDelightQueriesFile,
-    output: FileAppender
+    output: FileAppender,
+    generateAsync: Boolean,
```
instead of passing this around I would just pull it from the dialect, similar to [this](https://github.com/cashapp/sqldelight/blob/master/sqldelight-compiler/src/main/kotlin/app/cash/sqldelight/core/compiler/QueryGenerator.kt#L37)

AlecStrong replied:
This will also fix the currently failing intellij tests, and probably clean up some code

dellisd commented:
In terms of tests, many of the existing test suites can be duplicated and updated to check for `suspend` functions and `Async*` classes, but do you think it's worth duplicating _every_ test? Since most of the codgen logic is untouched I figure it could be enough to add tests just to cover the query function signatures for now.

AlecStrong commented:
yea i wouldn't worry about duplicating all tests, its fine to just test that specific behavior once and rely on the normal tests for everything else

dellisd commented:
I think this now covers the core of what's needed for async drivers at least in terms of codegen and runtime API.

The sqljs worker driver is incomplete (lacks prepared statements and transactions) but I think it might be better to leave that as a separate PR since it'll be pretty involved on its own. The R2DBC driver isn't very polished and has limited actual driver support, and as mentioned earlier it's kind of unclear what the demand for it is so I'm not sure if we even want to keep it right now? 

I'll add other small components like `Flow` extensions later (as a separate module from the existing extensions?)

But for now I think it's ready for some real reviews!

AlecStrong commented on test-util/src/main/kotlin/app/cash/sqldelight/test/util/FixtureCompiler.kt:
```diff
@@ -144,7 +148,7 @@ object FixtureCompiler {
         file!!,
         "testmodule",
-        fileWriter,
+        fileWriter
```
```suggestion
        fileWriter,
```

AlecStrong commented on dialects/sqlite-3-18/src/main/kotlin/app/cash/sqldelight/dialects/sqlite_3_18/SqliteDialect.kt:
```diff
@@ -21,6 +21,9 @@ import timber.log.Timber
  */
 open class SqliteDialect : SqlDelightDialect {
   override val driverType = ClassName("app.cash.sqldelight.db", "SqlDriver")
+  override val asyncDriverType: ClassName? = ClassName("app.cash.sqldelight.async.db", "AsyncSqlDriver")
+  override val asyncCursorType: ClassName? = ClassName("app.cash.sqldelight.async.db", "AsyncSqlCursor")
+  override val asyncPreparedStatementType: ClassName? = ClassName("app.cash.sqldelight.async.db", "AsyncSqlPreparedStatement")
```
Should these be promoted as default values and then some of the subclasses dont need these duplicated?

AlecStrong commented on sqldelight-gradle-plugin/src/test/integration-mysql-async/src/test/kotlin/app/cash/sqldelight/mysql/integration/MySqlTest.kt:
```diff
@@ -0,0 +1,90 @@
+    }
+  }
+
+  /* TODO: Dates are currently broken with the mysql r2dbc driver
```
whats the exception?

AlecStrong replied:
and i'd prefer just deleting the code over adding commented code

dellisd replied:
If I remember correctly the mysql driver can't return an `OffsetDateTime` which is expected. It looks like a future release of the driver will [return `ZonedDateTime` by default](https://github.com/mirromutth/r2dbc-mysql/pull/210) which I guess won't fix the issue, but it's also been almost two years since the last release.

I'll just remove this test and we can revisit it again if it comes up in the future.

AlecStrong commented on sqldelight-gradle-plugin/src/test/integration-mysql-async/src/test/kotlin/app/cash/sqldelight/mysql/integration/MySqlTest.kt:
```diff
@@ -0,0 +1,90 @@
+
+    return MyDatabase(driver).also { MyDatabase.Schema.create(driver) }
+  }
+
+  fun after() {
+  }
```
```suggestion
```

AlecStrong commented on sqldelight-gradle-plugin/src/test/integration-mysql-async/src/test/kotlin/app/cash/sqldelight/mysql/integration/MySqlTest.kt:
```diff
@@ -0,0 +1,90 @@
+    block(db)
+  }
+
+  suspend fun before(): MyDatabase {
+    val connection = factory.create().awaitSingle()
+    val driver = R2dbcDriver(connection)
+
+    return MyDatabase(driver).also { MyDatabase.Schema.create(driver) }
+  }
```
I would just inline this code into `runTest`

AlecStrong commented on sqldelight-gradle-plugin/src/main/kotlin/app/cash/sqldelight/gradle/SqlDelightPlugin.kt:
```diff
@@ -81,20 +82,24 @@ abstract class SqlDelightPlugin : Plugin<Project> {
 
+    val needsSyncRuntime = extension.databases.any { !it.generateAsync }
+    val needsAsyncRuntime = extension.databases.any { it.generateAsync }
+    val runtimeDependencies = mutableListOf<Dependency>().apply {
```
```suggestion
    val runtimeDependencies = buildList {
```

I think this is an API that exists

dellisd replied:
`java-gradle-plugin` depends on Kotlin 1.5.31 where that function is still experimental, otherwise I would have used it.

AlecStrong commented on sqldelight-gradle-plugin/src/main/kotlin/app/cash/sqldelight/gradle/SqlDelightDatabase.kt:
```diff
@@ -141,7 +142,8 @@ class SqlDelightDatabase(
         deriveSchemaFromMigrations = deriveSchemaFromMigrations,
-        treatNullAsUnknownForEquality = treatNullAsUnknownForEquality
+        treatNullAsUnknownForEquality = treatNullAsUnknownForEquality,
+        generateAsync = generateAsync
```
```suggestion
        generateAsync = generateAsync,
```

AlecStrong commented on sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/AsyncSelectQueryTypeTest.kt:
```diff
@@ -0,0 +1,154 @@
+      |RETURNING *;
+      |""".trimMargin(),
+      tempFolder,
+      dialect = PostgreSqlDialect(),
```
should use the burst parameter

AlecStrong replied:
or remove the burst parameter

dellisd replied:
This was copied straight from `SelectQueryTypeTest`. Do you want those ones changed in this PR too?

AlecStrong commented on sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/AsyncSelectQueryTypeTest.kt:
```diff
@@ -0,0 +1,154 @@
+      |RETURNING id, firstname, lastname;
+      |""".trimMargin(),
+      tempFolder,
+      dialect = PostgreSqlDialect(),
```
ditto

AlecStrong commented on sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/AsyncSelectQueryTypeTest.kt:
```diff
@@ -0,0 +1,154 @@
+
+    assertThat(generator.customResultTypeFunction().toString()).isEqualTo(
+      """
+      |public suspend fun <T : kotlin.Any> update(
```
do these functions need to be suspending? I guess it doesn't really matter since you almost always execute them in the same code block

dellisd replied:
Oh actually you're right, these shouldn't be suspending. The suspending call would be made on the returned `UpdateQuery`. For correctness-sake I'll remove the modifier there.

AlecStrong commented on sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/AsyncQueryFunctionTest.kt:
```diff
@@ -0,0 +1,47 @@
+package app.cash.sqldelight.core.queries
```
might be nice to plop all the async tests into a single folder

AlecStrong commented on sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/AsyncQueryFunctionTest.kt:
```diff
@@ -0,0 +1,47 @@
+import org.junit.rules.TemporaryFolder
+import org.junit.runner.RunWith
+
+@RunWith(BurstJUnit4::class)
```
```suggestion
```

AlecStrong commented on sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/AsyncMutatorQueryTypeTest.kt:
```diff
@@ -0,0 +1,141 @@
+import org.junit.rules.TemporaryFolder
+import org.junit.runner.RunWith
+
+@RunWith(BurstJUnit4::class)
```
```suggestion
```

AlecStrong commented on sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/AsyncTest.kt:
```diff
@@ -0,0 +1,214 @@
+import org.junit.rules.TemporaryFolder
+import java.io.File
+
+class AsyncTest {
```
I think we can probably drop this test, I trust that the other ones which cover selects/inserts/updates/etc will provide enough coverage

AlecStrong commented on sqldelight-compiler/src/main/kotlin/app/cash/sqldelight/core/compiler/QueryGenerator.kt:
```diff
@@ -256,13 +264,15 @@ abstract class QueryGenerator(
     } else if (optimisticLock != null) {
+      val execute = "val result = $DRIVER_NAME.execute"
       result.addStatement(
-        "val result = $DRIVER_NAME.execute($statementId, %P, %L)$binder",
+        "$execute($statementId, %P, %L)$binder",
         *arguments.toTypedArray()
       )
     } else {
+      val execute = "$DRIVER_NAME.execute"
       result.addStatement(
-        "$DRIVER_NAME.execute($statementId, %P, %L)$binder",
+        "$execute($statementId, %P, %L)$binder",
```
Can be reverted I think

AlecStrong commented on sqldelight-compiler/src/main/kotlin/app/cash/sqldelight/core/compiler/QueryGenerator.kt:
```diff
@@ -76,6 +78,11 @@ abstract class QueryGenerator(
     statement: PsiElement,
     id: Int
   ): CodeBlock {
+    if (generateAsync) {
+      check(dialect.asyncPreparedStatementType != null) { "Dialect $dialect does not support async drivers" }
+    }
```
If you promote the default types up then you dont need these checks, which I think is fine if by default dialects request the interface types.

AlecStrong replied:
I guess H2 is a problem because it would want to use APIs that dont exist in the interface type. FUN! I'd still promote it just to clean up the code, and we can worry about H2 when it pops up later.

AlecStrong commented on sqldelight-compiler/src/main/kotlin/app/cash/sqldelight/core/compiler/DatabaseGenerator.kt:
```diff
@@ -72,7 +78,10 @@ internal class DatabaseGenerator(
     // driver: SqlDriver
-    val dbParameter = ParameterSpec.builder(DRIVER_NAME, dialect.driverType).build()
+    if (generateAsync) {
+      check(dialect.asyncDriverType != null) { "Dialect $dialect does not support async drivers" }
```
when you've promoted the default type just search for these checks and remove them all

AlecStrong commented on sqldelight-compiler/dialect/src/main/kotlin/app/cash/sqldelight/dialect/api/SqlDelightDialect.kt:
```diff
@@ -9,16 +9,26 @@ interface SqlDelightDialect {
 
+  /**
+   * Dialect-specific implementation of an `AsyncSqlDriver`.
+   * TODO: Check if this needs to be nullable, also maybe combine these into a class
```
Had the same thought about sticking them in a class. `RuntimeTypes` or something like that. I think its a good idea

dellisd replied:
Do you want that in this PR? I guess may as well get it out of the way while adding the async types?

AlecStrong replied:
yea if it can be done in this PR I'd just do it now, I think it'll shrink some bits of it

AlecStrong reviewed (commented):
Worked my way through the codegen, will review the runtime bits tomorrow

AlecStrong commented on runtime/runtime-async/src/commonMain/kotlin/app/cash/sqldelight/async/db/Closeable.kt:
```diff
@@ -0,0 +1,31 @@
+  /**
+   * Close any resources backed by this object.
+   */
+  fun close()
```
yea i think thats probably right

dellisd commented:
I didn't realize how tightly-coupled the `driver-test` module is to SQLite so the async version of it (which is basically just a copy) may not be all that useful in its current form. e.g. Postgres has no `BLOB` type and no `changes()` function.

AlecStrong commented:
yea lets just remove it, we can revisit the existing driver test stuff along with potentially adding an async one later

AlecStrong commented on runtime/runtime-async/src/commonMain/kotlin/app/cash/sqldelight/async/AsyncTransacter.kt:
```diff
@@ -0,0 +1,267 @@
+  }
+
+  private suspend fun <R> transactionWithWrapper(noEnclosing: Boolean, wrapperBody: suspend AsyncTransactionWrapper<R>.() -> R): R {
+    val transaction = driver.newTransaction()
```
probably we should figure out how to share this logic between the sync/async transacter types, but we can do that in a follow up

AlecStrong commented on runtime/runtime-async/src/commonMain/kotlin/app/cash/sqldelight/async/logs/LogAsyncSqlDriver.kt:
```diff
@@ -0,0 +1,110 @@
+import app.cash.sqldelight.async.db.AsyncSqlDriver
+import app.cash.sqldelight.async.db.AsyncSqlPreparedStatement
+
+class LogAsyncSqlDriver(
```
lets just remove this from the async runtime, its not so valuable that I think we should start with it, and maybe theres a better event driven abstraction or something we can replace the normal runtime one with

dellisd replied:
The async runtime tests use it so I'll just move it to the test sources, otherwise we'd have zero tests.

AlecStrong commented on runtime/runtime/src/commonMain/kotlin/app/cash/sqldelight/Transacter.kt:
```diff
@@ -138,7 +133,7 @@ interface Transacter {
 }
 
-private class RollbackException(val value: Any? = null) : Throwable()
+internal class RollbackException(val value: Any? = null) : Throwable()
```
revert

AlecStrong reviewed (approved):
looks good, fantastic stuff

dellisd commented on dialects/hsql/src/main/kotlin/app/cash/sqldelight/dialects/hsql/HsqlDialect.kt:
```diff
@@ -13,9 +14,12 @@ import com.squareup.kotlinpoet.ClassName
  * Base dialect for JDBC implementations.
  */
 class HsqlDialect : SqlDelightDialect {
```
Ooh good idea

dellisd commented on drivers/r2dbc-driver/src/main/kotlin/app/cash/sqldelight/driver/r2dbc/R2dbcDriver.kt:
```diff
@@ -0,0 +1,195 @@
+      R2dbcPreparedStatement(statement).apply { if (binders != null) this.binders() }
+    }
+
+    val result = prepared.execute().awaitSingle()
+    // TODO: r2dbc-mysql emits a java.lang.Integer instead of a java.lang.Long, mysql driver needs to support latest r2dbc-spi
+    // return result.rowsUpdated.awaitSingle()
+    return 0L
```
An older version of the R2DBC SPI has the `execute()` method returning an `Int`, but in a newer version it was changed to a `Long`. The latest release of the MySQL driver implements the older version of the SPI so there's an ABI incompatibility that we can't do anything about.

I think I'm going to get rid of that comment and return the actual `rowsUpdated` result since this is a weird case to be handling, but we'll be left in a funny situation where there are zero R2DBC drivers that this SQLDelight driver can actually work with (as-is anyway). lol...

AlecStrong replied:
oh dear okay. Do you mean that `rowsUpdated` changed from an Int to a Long? Or do both old and new versions correctly support rowsUpdated so that would be fine

dellisd replied:
It changed from an Int to a Long, and the mysql driver just hasn't seen a new release in a long time (no pun intended)
